src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -83,6 +77,57 @@ void processRule(string name, ref string result)
                                 processRule(referencedRule, ref result);
                 }
             }
+
+            // Convert rules like `a: (x | y) ...` into:
+            //    a: x ...
+            //     | y ...;
+            bool trySplitTopLevelUnion(Node node)
+            {
+                if (node.Children[0] is Field field && field.Kinds.Count > 0)
+                {
+                    foreach (var kind in field.Kinds)
+                    {
+                        field.Kinds = new List<Kind> { kind };
+                        rules[node.Name].Add(HandleChildren(node.Children));
+                    }
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            // For nodes that have many subkinds (i.e. BinaryExpression) use the correspondance
+            // between node-kinds and child-kinds to generate N different productions corresponding
+            // to each kind.
+            bool trySplitCompoundNodes(Node node)
```

Comment:
this is new to match VB.

Reply:
I rolled this back.  i didn't think it significantly helped the generated c# grammar.  it just added a lot of named rules that were only referenced from the existing rule.  this is different than in VB.  in VB it causes us to generate rules that themselves can be referenced from other productions as well.